---
title: "Using the TcGSA package"
author: "Anthony Devaux"
date: "`r Sys.Date()`"
output:
  html_document:
    highlight: pygments
    number_sections: yes
    theme: united
    toc: yes
  pdf_document:
    toc: yes
  word_document:
    toc: yes
vignette: |
  %\VignetteIndexEntry{Vignette Title} %\VignetteEngine{knitr::rmarkdown} %\VignetteEncoding{UTF-8}
---

```{r pre, echo=FALSE, warning=FALSE, include=FALSE}
library(knitr)
knitr::opts_chunk$set(echo = TRUE, eval=TRUE, cache=TRUE)
```

# Overview of TcGSA

TcGSA (Time-course Gene Set Analysis *Hejblum et al., 2015*) package test the gene expression among the gene set and over the time. The gene set is made from the common biological utility of genes using linear mixed model to take into account the potential heterogeneity of expression within the gene sets. For more details, check the article [here](http://dx.plos.org/10.1371/journal.pcbi.1004310).

# Getting started using TcGSA

3 inputs are required to run `TcGSA` functions :

* The gene set object
* The gene expression matrix
* The design data matrix

## Gene set object

A gene set is a group of genes with the same biological function. It enables to detect different gene expression and seems to be more powerful than a gene-by-gene analysis. Several definitions of groups have been made, among the most important :

* Chaussabel's modules (*Chaussabel et al., 2008*)
* Gene Ontology database (*Ashburner et al., 2000*)
* Kyoto Encyclopedia of Genes and Genomes database (*Kanehisa et Goto, 2000*)

The gene set object is a `gmt` format containing :

* the name of the gene set
* the description of the gene set, such as biological function
* the list of probes representing the genes inside each gene set

2 methods can be used to build the gene set object, an easier or a hard.

### The easiest method

The easiest way to get the gene set object is to import it from the *Hejblum et al., 2015* supplementary file with this command : 

```{r GS_import, message=FALSE, cache=TRUE}
temp <- tempfile()
utils::download.file("https://doi.org/10.1371/journal.pcbi.1004310.s007", destfile = temp, mode = "wb")
load(unz(temp, "ReproducibleRFiles/GMTs_PLOScb.RData", open = "r"))
unlink(temp)
rm(temp)
```

This file contains 3 gene sets detailed above. But be careful with the version of the gene set database because they can be outdated.

To make sure to have the latest version of the database, you should create the `gmt` gene set object from the hard method below.

### The hardest method

To self-build the `gmt` gene set object, you have to prepare the `.gmt` file. This file format is the tab delimited file which can be created with this helpful website [here](http://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29). In this file, one row designs one gene set with :

* Column 1 : name of the gene set
* Column 2 : description of the gene set
* Remaining columns : list of genes included in the gene set (represented by the probes)

Next, to import the `.gmt` file into R, you need to run the `GSA.read.gmt` function from `GSA` package. More details on the `GSA` help package.

## Gene expression matrix

This matrix contains the gene expression (in cells) for each gene (in rows) of each sample (in columns) gathered by microarray analysis. The gene expression should be already normalized before using `TcGSA`. In rows, the name of each gene matched with the name of probes (representing genes) in the `gmt` gene set object. 

## Design data matrix

The design data matrix contains for each sample (in row), several variables (in column). The variables required for the matrix are :

* Name of samples
* Identifiant of patients
* Time measurements
* _In case of multiple treatments_, the name of treatment

Name of samples should be unique and match with the samples of gene expression matrix with the same order.

# How to use TcGSA for one treatment group ? 

## Data files

This example comes from *Hejblum et al., 2015* and DALIA-1 HIV therapeutic vaccine trial. The aim of this study is to evaluate the immune response to HIV vaccine. To conduct this study, 19 patients contaminated by the HIV have been followed for 48 weeks split into 2 phases of 24 each :

* 1st phase from week 0 to week 24 : the patients were under antiretroviral treatment and received trial vaccine on week 0, 4, 8 and 12.
* 2nd phase from week 24 to week 48 : for the follow-up of patients, none of trial vaccine has been injected and antiretroviral treatments have been interrupted (except for health problems).

Blood samples have been collected on the different time measurement for each patient to detect the variability of gene expression over time. For more details, check the article [here](http://dx.plos.org/10.1371/journal.pcbi.1004310).

The data are available on [GEO website](https://www.ncbi.nlm.nih.gov/geo/) with GEO access number 'GSE46734'. In this example, we are using the `GEOquery` package to get the data files from GEO website (see appendix for more details on `GEOquery`)

### Import of data files

In this example, we can see that the supplementary files are available on GEO website with GEO access number, so we can run the `getGEOSuppFiles` function to import data.

```{r import_dalia, message=FALSE, cache=TRUE}
GEOquery::getGEOSuppFiles('GSE46734')
```

### Design data matrix

The design data matrix (called `design_preATI`) is built from one of the GEO supplementary files. It contains the required variables plus some additional for this study. Some reprocessing are made according to the source of this paper.

```{r design_dalia, cache=TRUE}
design <- read.delim(gzfile("GSE46734/GSE46734_DALIA1longitudinalTranscriptome_DESIGN_anonym.txt.gz"))
design_preATI <- design[-which(design$TimePoint<0 | design$TimePoint==16 | design$TimePoint>22), ]
head(design_preATI,5)
```

Here we are :

* `Sample_name` for the name of samples
* `Patient_ID` for the identification of patients
* `TimePoint` for the time measurements
* `Chip_ID` `HYB_Chamber` `HYB_Day` are the variables not required for TcGSA commands

The entire matrix contains 90 samples and 6 variables
```{r dim_design_DALIA}
dim(design_preATI)
```

### Gene expression matrix

The gene expression matrix (called `expr_preATI`) is built from one of the GEO supplementary files. The data is already normalized.

```{r expr_dalia, cache=TRUE}
expr_preATI <- read.delim(gzfile("GSE46734/GSE46734_DALIA1longitudinalTranscriptome_PALO01_PreATI_NEQC_NonParamCombat.txt.gz"))
rownames(expr_preATI) <- expr_preATI$PROBE_ID
expr_preATI <- expr_preATI[,as.character(design_preATI$Sample_name)]

expr_preATI[1:4,1:4]
```

Here we are :

* In row `ILMN_xxxxxxx` for each gene of sample
* In column `Xxxxxxxxxxx_X` for the name of each sample
* In cell, gene expression

The entire matrix contains 32978 genes and 90 samples (number of samples should be the same as the design data matrix)
```{r dim_expr_DALIA}
dim(expr_preATI)
dim(design_preATI)
identical(dim(expr_preATI)[2],dim(design_preATI)[1])
```


## Likelihood ratios test

This function provides the result of likelihood ratio test using the linear mixed model for each gene set. For this example, we use gene sets data from Chaussabel's modules (*Chaussabel et al., 2008*). `TcGSA.LR` function requires :

* `expr` : name of the gene expression matrix
* `gmt` : name of the gmt gene set object
* `design` : name of the design data matrix
* `subject_name` : name of the identification of patients in the design data matrix
* `time_name` : name of the time measurements in the design data matrix


```{r LR_ST, message=FALSE, warning=FALSE, include=FALSE, cache=TRUE}
tcgsa_result <- TcGSA::TcGSA.LR(expr = expr_preATI, 
         gmt = gmt_modulesV2, 
         design = design_preATI, 
         subject_name = "Patient_ID", 
         time_name = "TimePoint")
```
```{r LR_ST_2, eval=FALSE, cache=TRUE}
tcgsa_result <- TcGSA::TcGSA.LR(expr = expr_preATI, 
         gmt = gmt_modulesV2, 
         design = design_preATI, 
         subject_name = "Patient_ID", 
         time_name = "TimePoint")
```

```{r tcgsa_result,cache=TRUE}
tcgsa_result
```

Now `tcgsa_result` is your TcGSA object containing the form of time trends (default option is `linear`), the number of treatment groups and the number of gene sets tested depending on which gene set base is defined. To get the number of significant gene sets, you can also use `summary` function on your TcGSA object.

```{r summary_dalia, cache=TRUE}
summary(tcgsa_result)
```

To get more details on the significant gene sets, use `signifLRT.TcGSA` function. It returns several information such as the significant gene sets among all the gene sets tested and their p-values with adjustment for multiple testing (defaut option `BY` for Benjamini-Yekutieli correction *Benjamini et Yekutieli, 2001*) and 5% threshold. Here, an example of five significant gene sets.

```{r signifLRT_ST, cache=TRUE}
TcGSA::signifLRT.TcGSA(tcgsa_result)$mixedLRTadjRes[1:5,]
```

You can also use `multtest.TcGSA` function to provide the likelihood ratios, the raw and adjusted p-values for the whole gene sets with 5% threshold. Here, an example of five results :

```{r multtest_ST, cache=TRUE}
TcGSA::multtest.TcGSA(tcgsa_result)[1:5,]
```

`CVG_H0` and `CVG_H1` are the convergence of the model under null and alternative hypotheses. Close to 0 means a good convergence of the model.

## Graphical output

`plot1GS` plots the different representations of gene expression in a specific gene set defined by `geneset.name` argument. `plot1GS` function requires :

* `expr` : either the name of the gene expression matrix or the estimations of linear mixed model. In this example, we used the raw data from the gene expression matrix.
* `gmt` : name of the gmt gene set object
* `Subject_ID` : name of the identification of patients in the design data matrix
* `TimePoint` : name of the time measurements of the design data matrix
* `geneset.name` : name of gene set. The significant can be found with `signifLRT.TcGSA(tcgsa_result)$mixedLRTadjRes`
* `time_unit` : value be displayed next to the values of `TimePoint` on the x-axis __(OPTIONAL)__


```{r plot1GS_ST, message=FALSE, warning=FALSE, fig.keep='all',cache=TRUE}
TcGSA::plot1GS(expr = expr_preATI, 
#plot1GS(expr = tcgsa_result$Estimations,
        gmt = gmt_modulesV2, 
        Subject_ID = design_preATI$Patient_ID, 
        TimePoint = design_preATI$TimePoint,
        clustering = FALSE, 
        time_unit = "W", 
        geneset.name = "M3.2", 
        title="",
        margins=0.4,
        lab.cex=0.37,
        axis.cex=0.37,
        line.size=0.45,
        gg.add=list(theme(legend.position="none"),
                    ylim(-1.26,1.26)
        ) )
```

Dotted line shows the trend(s) of the gene set over time.

# How to use TcGSA for several treatment group ?

This example comes from *Obermoser et al. 2013* to study the responses to influenza and pneumococcal vaccines on healthy patients using gene expression over time.
The patients are split into 3 groups of 6, receiving either 2009-2010 seasonal influenza vaccine (Fluzone), a 23-valent pneumococcal vaccine (Pneumovax23) or a placebo.  
Blood samples have been taken on a day -7, 0, 1, 3, 7, 10, 14, 21 and 28 to analyze gene expression over time. For more details, check the article [here](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3492754/).

The data are available on [GEO website](https://www.ncbi.nlm.nih.gov/geo/) with GEO access number 'GSE30101'. In this example, we are using the `GEOquery` package to get the data files from GEO website (see appendix for more details on `GEOquery`)

## Data preparation

In this example, we can see that the supplementary files are available on GEO website with GEO access number, so we can run the `getGEOSuppFiles` function to import data.

In this case, the design and expression data matrix are not available on the GEO website using GEO access number. We have to download the raw file and import with `getGEO` function.

```{r import_ober, message=FALSE, warning=FALSE, cache=TRUE}
download.file("ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE30nnn/GSE30101/soft/GSE30101_family.soft.gz", destfile = "GSE30101_family.soft.gz", mode = "wb", cacheOK = FALSE)
gse.soft <- GEOquery::getGEO(filename="GSE30101_family.soft.gz")
```

In opposite to the first example, more processing should be done on this data to build the design and gene expression data.

Gene expression matrix can be made from GEO file with these commands :

```{r expr_ober, cache=TRUE}
probesIDs <- GEOquery::Table(GEOquery::GSMList(gse.soft)[[1]])$ID
data.matrix <- do.call('cbind',lapply(GEOquery::GSMList(gse.soft),function(x)
                                      {tab <- Table(x)
                                       mymatch <- match(probesIDs,tab$ID_REF)
                                       return(tab$VALUE[mymatch])
                                     }))
rownames(data.matrix) <- probesIDs
expr.All.ChaussVac <- apply(X=data.matrix, FUN=as.numeric, MARGIN=2)
rownames(expr.All.ChaussVac) <- probesIDs
```

Design data matrix can be made from GEO file with these commands :

```{r design_ober, cache=TRUE}
design_list <-lapply(GEOquery::GSMList(gse.soft), function(x){Meta(x)$characteristics_ch1})
design <- data.frame(row.names=names(design_list))
design$sample_ID <- names(design_list)
s_id <- unlist(lapply(design_list, function(x){
  gsub("subject id: ", "", x[grep("subject id: ", x)])
}))
design$Subject_ID <- as.character(paste("P", s_id[design$sample_ID], sep=""))

time <- unlist(lapply(design_list, function(x){
  gsub("day: ", "", x[grep("day: ", x)])
}))
time[which(time %in% c("-7", "0.5", "1", "7", "10", "14",  "21", "28"))] <-
  paste("D", time[which(time %in% c("-7", "0.5", "1", "7", "10", "14", "21", "28"))], sep="")
time[which(time %in% c("-168", "1.5", "6", "9", "12", "15", "24", "36", "48"))] <-
  paste("H", time[which(time %in% c("-168", "1.5", "6", "9", "12", "15", "24", "36", "48"))], sep="")
design$Time <- as.character(time[design$sample_ID])

vac <- unlist(lapply(design_list, function(x){
  gsub("vaccine: ", "", x[grep("vaccine: ", x)])
}))
vac <- as.factor(vac)
levels(vac) <- c("influenza", "influenza", "influenza", "influenza", "saline", "pneumo", "pneumo", "pneumo", "saline", "saline")
design$Vaccine <- as.character(vac[design$sample_ID])

sampSet <- unlist(lapply(design_list, function(x){
  gsub("sample set: ", "", x[grep("sample set: ", x)])
}))
design$sampSet <- as.character(sampSet[design$sample_ID])

design$Time[which(design$sampSet=="Training_Set_Vein" & design$Time %in% c("0", "3"))] <-
  paste("D", design$Time[which(design$sampSet=="Training_Set_Vein" & design$Time %in% c("0", "3"))], sep="")
design$Time[which(design$sampSet=="Training_Set_Finger" & design$Time %in% c("0", "3"))] <-
  paste("H", design$Time[which(design$sampSet=="Training_Set_Finger" & design$Time %in% c("0", "3"))], sep="")
design$Time[which(design$sampSet=="Test_Set_Vein" & design$Time %in% c("0", "3"))] <-
  paste("D", design$Time[which(design$sampSet=="Test_Set_Vein" & design$Time %in% c("0", "3"))], sep="")
design$Time[which(design$sampSet=="Test_Set_Finger" & design$Time %in% c("0", "3"))] <-
  paste("D", design$Time[which(design$sampSet=="Test_Set_Finger" & design$Time %in% c("0", "3"))], sep="")
design$Time[which(design$sampSet=="Validation_Vein" & design$Time %in% c("0", "3"))] <-
  paste("D", design$Time[which(design$sampSet=="Validation_Vein" & design$Time %in% c("0", "3"))], sep="")

design$Day <- gsub("D", "", design$Time)
design$Day[grep("H", design$Day)] <- as.numeric(gsub("H", "", design$Day[grep("H", design$Day)]))/24
design$Day <- as.numeric(design$Day)

design.All.ChaussVac <- design



# Avg Baseline -----
design.All.ChaussVac.trainSetVein <- design.All.ChaussVac[which(design.All.ChaussVac$sampSet=="Training_Set_Vein"),]
samplesSaline2rmv <- design.All.ChaussVac.trainSetVein[162:214,"sample_ID"]
design.All.ChaussVac.noDup <- design.All.ChaussVac.trainSetVein[-which(design.All.ChaussVac.trainSetVein$sample_ID%in%samplesSaline2rmv),]

design.All.ChaussVac.AvgBl <- design.All.ChaussVac.noDup[which(design.All.ChaussVac.noDup$Day!=0),]
design.All.ChaussVac.AvgBl[which(design.All.ChaussVac.AvgBl$Day==-7),"Day"] <- 0
design.All.ChaussVac.AvgBl[which(design.All.ChaussVac.AvgBl$Time=="D-7"),"Time"] <- "D0"

expr.All.ChaussVac.AvgBl <- expr.All.ChaussVac[, design.All.ChaussVac.AvgBl$sample_ID]
for(p in unique(design.All.ChaussVac.AvgBl$Subject_ID)){
  if(length(which(design.All.ChaussVac.noDup$Subject_ID==p & (design.All.ChaussVac.noDup$Day==0 | design.All.ChaussVac.noDup$Day==-7)))>1){
    expr.All.ChaussVac.AvgBl[, which(design.All.ChaussVac.AvgBl$Subject_ID==p & design.All.ChaussVac.AvgBl$Day==0)] <-
      apply(X=cbind(expr.All.ChaussVac[, design.All.ChaussVac.noDup[which(design.All.ChaussVac.noDup$Subject_ID==p & design.All.ChaussVac.noDup$Day==0), "sample_ID"]],
                    expr.All.ChaussVac[, design.All.ChaussVac.noDup[which(design.All.ChaussVac.noDup$Subject_ID==p & design.All.ChaussVac.noDup$Day==-7), "sample_ID"]]),
            MARGIN=1, FUN=mean, na.rm=TRUE
            )
  }
}
rownames(expr.All.ChaussVac.AvgBl) <- probesIDs

if(!all.equal(as.character(design.All.ChaussVac.AvgBl$sample_ID), colnames(expr.All.ChaussVac.AvgBl))){stop("\n\n\nWARNING: EXPRESSION FILE ORDER NOT MATCHING DESIGN FILE\n\n\n")}
design.All.ChaussVac.AvgBl$Subject_ID <- as.factor(design.All.ChaussVac.AvgBl$Subject_ID)


design.PNEUMOvsSALINE.ChaussVac.AvgBl <- design.All.ChaussVac.AvgBl[which(design.All.ChaussVac.AvgBl$Vaccine!="influenza"), ]
design.PNEUMOvsSALINE.ChaussVac.AvgBl$Vaccine <- as.factor(as.character(design.PNEUMOvsSALINE.ChaussVac.AvgBl$Vaccine))

expr.PNEUMOvsSALINE.ChaussVac.AvgBl <- expr.All.ChaussVac.AvgBl[,design.PNEUMOvsSALINE.ChaussVac.AvgBl$sample_ID]

```


## Likelihood ratios test

This function provides the result of likelihood ratio test using the linear mixed model for each gene set. For this example, we use gene sets data from Chaussabel's modules (*Chaussabel et al., 2008*). `TcGSA.LR` function requires :

* `expr` : the gene expression matrix
* `gmt` : the gmt gene set object
* `design` : the design data matrix
* `subject_name` : the identification of patients in the design data matrix
* `time_name` : the time measurements in the design data matrix
* `group_name` : the group of treatment in the design data matrix

```{r LR_MT, message=FALSE, warning=FALSE, include=FALSE, cache=TRUE}
tcgsa_result_MT <- TcGSA::TcGSA.LR(expr = expr.PNEUMOvsSALINE.ChaussVac.AvgBl, 
         gmt = gmt_modulesV2,
         design = design.PNEUMOvsSALINE.ChaussVac.AvgBl, 
         subject_name = "Subject_ID", 
         time_name = "Day", 
         group_name = "Vaccine")
```
```{r LR_MT_2, eval=FALSE, cache=TRUE}
tcgsa_result_MT <- TcGSA::TcGSA.LR(expr = expr.PNEUMOvsSALINE.ChaussVac.AvgBl, 
         gmt = gmt_modulesV2,
         design = design.PNEUMOvsSALINE.ChaussVac.AvgBl, 
         subject_name = "Subject_ID", 
         time_name = "Day", 
         group_name = "Vaccine")
```

```{r tcgsa_result_MT, cache=TRUE}
tcgsa_result_MT
```

Now `tcgsa_result_MT` is your TcGSA object containing the form of time trends (default option is `linear`), the number of treatment groups and the number of gene sets tested depending on which gene set base is defined. To get the number of significant gene set, you can also use `summary` function on your TcGSA object.

```{r summary_ober, cache=TRUE}
summary(tcgsa_result_MT)
```

To get more details on the significant gene sets, use `signifLRT.TcGSA` function. It returns several information such as the significant gene sets among all the gene sets tested and their p-values with adjustment for multiple testing (default option `BY` for Benjamini-Yekutieli correction *Benjamini et Yekutieli, 2001*) and 5% threshold. Here, an example of five significant gene sets.

```{r signifLRT_MT, cache=TRUE}
TcGSA::signifLRT.TcGSA(tcgsa_result_MT)$mixedLRTadjRes[1:5,]
```

You can also use `multtest.TcGSA` function to provide the likelihood ratios, the raw and adjusted p-values for the whole gene sets with 5% threshold. Here, an example of five results :

```{r multtest_MT, cache=TRUE}
TcGSA::multtest.TcGSA(tcgsa_result_MT)[1:5,]
```

`CVG_H0` and `CVG_H1` are the convergence of the model under null and alternative hypotheses. Close to 0 means a good convergence of the model.

## Heatmap of significant gene sets

### Make clusters from tcgsa object

`clustTrend` builds clusters of genes from their trends dynamics. `clustTrend` function requires :

* `tcgs` : your TcGSA object
* `expr` : estimation of gene expressions with linear mixed model from TcGSA object
* `Subject_ID` : name of the identification of patients in the design data matrix
* `TimePoint` : name of the time measurements in the design data matrix
* `baseline` _OPTIONAL_ : value of `TimePoint` used to be the reference
* `group_of_interest` : name of a treatment in the design data matrix


```{r clust_MT, message=FALSE, warning=FALSE, include=FALSE, cache=TRUE}
clust <- TcGSA::clustTrend(tcgs = tcgsa_result_MT, 
                    expr=tcgsa_result_MT$Estimations,
                    Subject_ID=design_gen$Patient_ID,
                    TimePoint=design_gen$TimePoint,
                    baseline = 0,
                    group_of_interest="pneumo")
```
```{r clust_MT_2, eval=FALSE, cache=TRUE}
clust <- TcGSA::clustTrend(tcgs = tcgsa_result_MT, 
                    expr=tcgsa_result_MT$Estimations,
                    Subject_ID=design_gen$Patient_ID,
                    TimePoint=design_gen$TimePoint,
                    baseline = 0,
                    group_of_interest="pneumo")
```

```{r clust_MT_3}
clust
```

`clust` shows the number of trends within the significant gene sets.

### Plot heatmap

`plot` draws different kinds of graphics, but we focus on heatmap graphics. This function requires :

* `x` : TcGSA object
* `expr` : estimation of gene expressions with linear mixed model from TcGSA object
* `Subject_ID` : name of the identification of patients in the design data matrix
* `TimePoint` : name of the time measurements of the design data matrix
* `group_of_interest` : name of a treatment in the design data matrix
* `clust_trends` : cluster object with the clusters of genes from their trends dynamics


```{r heatmap_MT, message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}
plot(x=tcgsa_result_MT, expr=tcgsa_result_MT$Estimations,
     Subject_ID=design_gen$Patient_ID,
     TimePoint=design_gen$TimePoint,
     group_of_interest="pneumo",
     clust_trends=clust,
     legend.breaks=seq(from=-2,to=2, by=0.01), time_unit="D",
     subtitle="Pneumo vs Saline", cex.label.row=2.5, cex.label.col=3,
     heatmap.width=0.4, dendrogram.size=0.25, margins=c(5,23),
     heatKey.size=0.13)
```

The heatmap shows an under (blue color) or an over (red color) gene expression for each significant gene sets compared to `baseline` value (which represents the time measurement reference) in `clust` object. The clusters are plotted with the dendrogram

# References

Ashburner M, Ball CA, Blake JA, Botstein D, Butler H, Cherry JM, et al., (2000) Gene Ontology: tool for the unification of biology. _Nat Genet_ 25(1):25-9. 

Benjamini Y, Yekutieli D, (2001) The Control of the False Discovery Rate in Multiple Testing under Dependency. _Ann Stat_ 29(4):1165-88. 

Chaussabel D, Quinn C, Shen J, Patel P, Glaser C, Baldwin N, et al., (2008) A Modular Analysis Framework for Blood Genomics Studies: Application to Systemic Lupus Erythematosus. _Immunity_ 29(1):150-64. 

Hejblum BP, Skinner J, Thiebaut R, (2015) Time-Course Gene Set Analysis for Longitudinal Gene Expression Data. _PLOS Comput Biol_ 11(6):e1004310.

Kanehisa M, Goto S, (2000) KEGG: Kyoto Encyclopedia of Genes and Genomes. _Nucleic Acids Res_ 28(1):27-30.

Obermoser G, Presnell S, Domico K, Xu H, Wang Y, Anguiano E, et al., (2013) Systems Scale Interactive Exploration Reveals Quantitative and Qualitative Differences in Response to Influenza and Pneumococcal Vaccines. _Immunity_ 38(4):831-44. 

# Appendix

## GEOquery package

In case of your data are public and available on Gene Expression Omnibus (GEO), the public database with gene expression data, you have to download GEOquery package with these commands :

```{r dl_GEOquery, warning=FALSE, message=FALSE, cache=TRUE}
## try http:// if https:// URLs are not supported
source("https://bioconductor.org/biocLite.R")
biocLite("GEOquery")
```

Then you could import them from [GEO website](https://www.ncbi.nlm.nih.gov/geo/) with the GEO access number (build like this : _GSExxxxx_) and run the commands with this number. 
`getGEOSuppFiles` provides several files including the gene expression file and design data file and don't need to reprocess them. Sometimes, theses files are not available on GEO website. In this case, you have to run the `getGEO` function from the file containing the raw data. Moreover, you need some reprocessing to build the gene expression, design data matrix. An example for each function is available on "How to use TcGSA for one treatment group ?" and "How to use TcGSA for several treatment group ?" parts.

```{r library, warning=FALSE, message=FALSE, cache=TRUE}
library(TcGSA)
library(GEOquery)
```

More details on the main article : Davis S, Meltzer P, (2007) GEOquery: a bridge between the Gene Expression Omnibus (GEO) and BioConductor _Bioinformatics_ 14:1846-1847.